pkg/kubelet/preemption/preemption_test.go

```diff
@@ -0,0 +1,480 @@
+			inputPods:             []*v1.Pod{allPods[bestEffort]},
+			insufficientResources: getAdmissionRequirementList(0, 1, 0),
+			expectErr:             true,
+			expectedOutput:        nil,
+		},
+		{
+			testName: "multiple pods evicted",
+			inputPods: []*v1.Pod{
+				allPods[critical], allPods[bestEffort], allPods[burstable], allPods[highRequestBurstable],
+				allPods[guaranteed], allPods[highRequestGuaranteed]},
+			insufficientResources: getAdmissionRequirementList(0, 550, 0),
+			expectErr:             false,
+			expectedOutput:        []*v1.Pod{allPods[highRequestBurstable], allPods[highRequestGuaranteed]},
+		},
+	}
+	for _, r := range runs {
+		podProvider.setPods(r.inputPods)
+		outErr := criticalPodAdmissionHandler.evictPodsToFreeRequests(r.insufficientResources)
+		outputPods := podKiller.getKilledPods()
+		if !r.expectErr && outErr != nil {
+			t.Errorf("evictPodsToFreeRequests returned an unexpected error during the %s test.  Err: %v", r.testName, outErr)
+		} else if r.expectErr && outErr == nil {
+			t.Errorf("evictPodsToFreeRequests expected an error but returned a successful output=%v during the %s test.", outputPods, r.testName)
+		} else if !podListEqual(r.expectedOutput, outputPods) {
+			t.Errorf("evictPodsToFreeRequests expected %v but got %v during the %s test.", r.expectedOutput, outputPods, r.testName)
+		}
+		podKiller.clear()
+	}
+}
+
+func BenchmarkGetPodsToPreempt(t *testing.B) {
+	allPods := getTestPods()
+	inputPods := []*v1.Pod{}
+	for i := 0; i < maxPods; i++ {
+		inputPods = append(inputPods, allPods[tinyBurstable])
+	}
+	for n := 0; n < t.N; n++ {
+		getPodsToPreempt(inputPods, admissionRequirementList([]*admissionRequirement{
+			{
+				resourceName: v1.ResourceCPU,
+				quantity:     parseCPUToInt64("110m"),
+			}}))
+	}
+}
+
+func TestGetPodsToPreempt(t *testing.T) {
+	type testRun struct {
+		testName              string
+		inputPods             []*v1.Pod
+		insufficientResources admissionRequirementList
+		expectErr             bool
+		expectedOutput        []*v1.Pod
+	}
+	allPods := getTestPods()
+	runs := []testRun{
+		{
+			testName:              "no requirements",
+			inputPods:             []*v1.Pod{},
+			insufficientResources: getAdmissionRequirementList(0, 0, 0),
+			expectErr:             false,
+			expectedOutput:        []*v1.Pod{},
+		},
+		{
+			testName:              "no pods",
+			inputPods:             []*v1.Pod{},
+			insufficientResources: getAdmissionRequirementList(0, 0, 1),
+			expectErr:             true,
+			expectedOutput:        nil,
+		},
+		{
+			testName:              "equal pods and resources requirements",
+			inputPods:             []*v1.Pod{allPods[burstable]},
+			insufficientResources: getAdmissionRequirementList(100, 100, 1),
+			expectErr:             false,
+			expectedOutput:        []*v1.Pod{allPods[burstable]},
+		},
+		{
+			testName:              "higer requirements than pod requests",
```

Comment:
This can happen only if pod requests are greater than allocatable. We should fail fast in such scenarios rather than going through the preemption logic.